Block Supports: Allow skipping serialization of individual features

aaronrobertshaw opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/pull/36104

## Description

This PR allows blocks to skip the serialization of individual features within a block support toolset. For example, only skipping the serialization of `text-decoration` styles for the typography support.

Updates include:
- Modifying the `addSaveProps` filter from the `style.js` hook
- Adding some tests for the skip serialization logic
- Updating the individual block support hooks to handle array values for skip serialization flag
- Updating application of link color and block gap styles to take into account block serialization config

## How has this been tested?

1. Ran style hook tests `npm run test-unit packages/block-editor/src/hooks/test/style.js`
2. Tested static blocks:
    - Checked block supports still function
    - Ensured skipping serialization via the current boolean `__experimentalSkipSerialization` flag still works
    - Tested skipping individual features via an array configuration for `__experimentalSkipSerialization`
3. Repeated the above checks but for dynamic blocks
4. Ran PHP unit tests: 

```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/block-supports/spacing-test.php
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/block-supports/colors-test.php
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/block-supports/border-test.php
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/block-supports/typography-test.php
```

### Testing Summary

You can use `group/block.json` as the basis for tests. Test using both the boolean – `"__experimentalSkipSerialization": true` – and the individual features, e.g., 
<details><summary>Individual skip serialization example for Group block.json</summary>

```json
		"color": {
			"gradients": true,
			"link": true,
			"__experimentalDefaultControls": {
				"background": true,
				"text": true
			},
			"__experimentalSkipSerialization": [ "gradients", "link", "background", "text" ]
		},
		"spacing": {
			"margin": [ "top", "bottom" ],
			"padding": true,
			"blockGap": true,
			"__experimentalDefaultControls": {
				"padding": true,
				"blockGap": true
			},
			"__experimentalSkipSerialization": [ "blockGap", "margin", "padding" ]
		},
		"__experimentalBorder": {
			"color": true,
			"radius": true,
			"style": true,
			"width": true,
			"__experimentalDefaultControls": {
				"color": true,
				"radius": true,
				"style": true,
				"width": true
			},
			"__experimentalSkipSerialization": [ "color", "radius", "style", "width" ]
		},
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalLetterSpacing": true,
			"__experimentalTextTransform": true,
			"__experimentalDefaultControls": {
				"fontSize": true
			},
			"__experimentalSkipSerialization": [ "fontSize", "lineHeight", "fontStyle", "fontWeight", "letterSpacing", "textTransform" ]
		},
```
</details>

#### Group Block (static block)

| Support | Feature    | Works | Boolean Skip | Array Skip |
| ------- | ---------- |:-----:|:------------:|:----------:|
| Color   | Text       |  ✅   |      ✅      |     ✅     |
| Color   | Background |  ✅   |      ✅      |     ✅     |
| Color   | Gradient   |  ✅   |      ✅      |     ✅     |
| Color   | Link       |  ✅   |      ✅      |     ✅     |
| Border  | Color      |  ✅   |      ✅      |     ✅     |
| Border  | Radius     |  ✅   |      ✅      |     ✅     |
| Border  | Style      |  ✅   |      ✅      |     ✅     |
| Border  | Width      |  ✅   |      ✅      |     ✅     |
| Spacing | Padding    |  ✅   |      ✅      |     ✅     | 


#### Buttons Block (static block)

| Support | Feature   | Works | Boolean Skip | Array Skip |
| ------- | --------- |:-----:|:------------:|:----------:|
| Spacing | Margin    |  ✅   |      ✅      |     ✅     |
| Spacing | Block Gap |  ✅   |      ✅      |     ✅     | 

#### Paragraph Block (static block)

| Support    | Feature        | Works | Boolean Skip | Array Skip |
| ---------- | -------------- |:-----:|:------------:|:----------:|
| Typography | Font Size      |  ✅   |      ✅      |     ✅     |
| Typography | Line Height    |  ✅   |      ✅      |     ✅     |
| Typography | Font Style     |  ✅   |      ✅      |     ✅     |
| Typography | Font Weight    |  ✅   |      ✅      |     ✅     |
| Typography | Letter Spacing |  ✅   |      ✅      |     ✅     |
| Typography | Text Transform |  ✅   |      ✅      |     ✅     | 

The following block supports were manually added for testing purposes.

| Support    | Feature         | Works | Boolean Skip | Array Skip |
| ---------- | --------------- |:-----:|:------------:|:----------:|
| Typography | Font Family     |  ✅   |      ✅      |     ✅     |
| Typography | Text Decoration |  ✅   |      ✅      |     ✅     | 

#### Site Title ( dynamic block )

| Support    | Feature        | Works | Boolean Skip | Array Skip |
| ---------- | -------------- |:-----:|:------------:|:----------:|
| Color      | Text           |  ✅   |      ✅      |     ✅     |
| Color      | Background     |  ✅   |      ✅      |     ✅     |
| Color      | Gradient       |  ✅   |      ✅      |     ✅     |
| Color      | Link           |  ✅   |      ✅      |     ✅     |
| Typography | Font Family    |  ✅   |      ✅      |     ✅     |
| Typography | Font Size      |  ✅   |      ✅      |     ✅     |
| Typography | Line Height    |  ✅   |      ✅      |     ✅     |
| Typography | Font Style     |  ✅   |      ✅      |     ✅     |
| Typography | Font Weight    |  ✅   |      ✅      |     ✅     |
| Typography | Letter Spacing |  ✅   |      ✅      |     ✅     |
| Typography | Text Transform |  ✅   |      ✅      |     ✅     | 
| Spacing    | Padding        |  ✅   |      ✅      |     ✅     |
| Spacing    | Margin         |  ✅   |      ✅      |     ✅     |

The following block supports were manually added for testing purposes.

| Support | Feature | Works | Boolean Skip | Array Skip |
| ------- | ------- |:-----:|:------------:|:----------:|
| Border  | Color   |  ✅   |      ✅      |     ✅     |
| Border  | Radius  |  ✅   |      ✅      |     ✅     |
| Border  | Style   |  ✅   |      ✅      |     ✅     |
| Border  | Width   |  ✅   |      ✅      |     ✅     | 


#### Navigation ( dynamic block )

| Support    | Feature         | Works | Boolean Skip | Array Skip |
| ---------- | --------------- |:-----:|:------------:|:----------:|
| Spacing    | Block Gap       |  ✅   |      ✅      |     ✅     |
| Typography | Text Decoration |  ✅   |      ✅      |     ✅     |


## Types of changes

Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Update styles hook to allow skipping individual feature serialization

aaronrobertshaw committed:
Fix typo

aaronrobertshaw committed:
Update JS serialization checks

aaronrobertshaw committed:
Update PHP block support serialization checks

aaronrobertshaw committed:
Add skip serialization checks for link colors and block gap

aaronrobertshaw committed:
Skip enforcing inline color styles and fix gradients

aaronrobertshaw commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -181,6 +202,19 @@ const skipSerializationPathsSave = {
+ * @constant
+ * @type {Record<string, string>}
+ */
+const renamedFeatures = { gradients: 'gradient' };
```
All but `color.gradients` map neatly between the support flag and the property within the style object. In the case of `gradients` the property that needs to be omitted from the styles is `gradient`. 

I ended up going with the option here of explicitly mapping this. We can therefore keep the flag to opt in to support the same as the value needed to skip serialization of just the `gradient` feature without having to migrate the support flag or style object property etc.

If there are better options, I'll happily adopt them.

ramonjd replied:
I think this is nice way forward for the sake of an outlying plural 😄 and to keep the block.json keys predictable and consistent.

Having the same key for `color.gradients` to `color.__experimentalSkipSerialization.gradients`, in my opinion, is worth it:

```json
		"color": {
			"gradients": true,
			"__experimentalSkipSerialization": [ "gradients" ],
                        ...
```

ramonjd commented:
This is working well for me so far. 

I've been testing using the Group block.

Skipping serialization for individual properties, e.g.,

```json
		"__experimentalBorder": {
			"color": true,
			"radius": true,
			"style": true,
			"width": true,
			"__experimentalSkipSerialization": [ "style" ],
                         ...
```

isolates that property, and it's no longer available in the block's `props` object via `useBlockProps()`

An empty array – `"__experimentalSkipSerialization": [],` – is equivalent to `"__experimentalSkipSerialization": false`, as I'd expect.

I've rebased to kick off the e2es again, and will keep testing the individual properties. 🍺

ramonjd committed:
The call to `gutenberg_skip_color_serialization()` from elements.php sometimes does not pass a valid $block_type, which must be an object. It is `NULL`. Perhaps due to a race condition. This commit checks if $block_type is an object before trying to use it.

ramonjd commented:
We were getting a PHP warning –

```
Notice: Trying to get property 'supports' of non-object in /var/www/html/wp-content/plugins/gutenberg/lib/block-supports/colors.php on line 158
```

– because `$block_type` passed to `gutenberg_skip_color_serialization( $block_type, 'link' );` in [elements.php]( https://github.com/WordPress/gutenberg/blob/fb39c77f4ad19b53a2b3d75f910fa8d24e8c3492/lib/block-supports/elements.php#L22) was `NULL` . 

5d99fb78daaf75b6a0fabd3adf0b0de35a83bf3e checks `$block_type` before using it.

ramonjd commented:
I think `blockGap` isn't working since it's controlled via the layout abstraction and there doesn't appear to be any skip serialization tests there.

_Edit_: 7bdb4113c40f3430a9de10acd55ca60d53da2a46 addresses this by ensuring that the `blockGap` value, where defined by the user in the editor, is not applied in the inline styles in the editor or on the frontend.

ramonjd committed:
This commit ensures that defining `"__experimentalSkipSerialization": [ "blockGap" ]` has an effect on the frontend and in the editor.
The blockGap value, where defined by the user, will not be applied.

Added a comment clarifying the use of skip serialization in relation to blockGap

ramonjd commented:
The fixture failures are curious. They started failing after 7bdb4113c40f3430a9de10acd55ca60d53da2a46. Reverting the commit locally via GIT makes them pass again.

**But** also locally, when I roll back the changes individually in each file (basically reverting manually), the tests still fail.

The failures themselves are related to the order of block attributes only.

🤔

andrewserong commented:
> The failures themselves related to the order of block attributes only.

@ramonjd just taking a look at [the commit](https://github.com/WordPress/gutenberg/commit/7bdb4113c40f3430a9de10acd55ca60d53da2a46) where the fixture tests started to fail. In that change, `shouldSkipSerialization` is now being imported from `../hooks/style.js` — however that file has a side-effect in that at the end of the file it registers a bunch of `addFilter` calls against `addAttribute`, `addSaveProps`, etc. So, if those filters are being registered in a different order / at a different time — could that then mean that the attributes are being added to the attributes object in a different sequence, which could then cause serialization to save out the block attributes in a different order?

One thing to try might be to move `shouldSkipSerialization` to a utility file that doesn't have side effects when imported, to see if that fixes the issue (or if it winds up causing other issues 😅!)

ramonjd commented:
> could that then mean that the attributes are being added to the attributes object in a different sequence, which could then cause serialization to save out the block attributes in a different order?

Definitely an avenue for investigation. Much appreciated @andrewserong I'll look into it.

What was baking my noodle was that even when I removed that import, and made all changed files identical with trunk (not via a git reset to the previous commit)… they still failed.

🥴

andrewserong committed:
Move shouldSkipSerialization to utils module to avoid side-effects of importing from style.js

andrewserong committed:
Format whitespace of an import

ramonjd committed:
Updated tests to test the `renamedFeatures` object.

ramonjd committed:
Adding block supports tests
Adding skip serialization tests

ramonjd commented:
I've rebased and retested all the skipped block support properties 😓 I think this is ready for another review.

glendaviesnz commented on lib/block-supports/colors.php:
```diff
@@ -142,6 +141,31 @@ function gutenberg_apply_colors_support( $block_type, $block_attributes ) {
+ * @return boolean Whether to serialize color support styles & classes.
+ */
+function gutenberg_skip_color_serialization( $block_type, $feature = null ) {
+	if ( ! is_object( $block_type ) ) {
```
The other skip serialization methods don't have this guard? Rather than add it to all of them is it possible to make this a shared method and pass in `$path` as that seems to be the only difference between them?

ramonjd replied:
Good point @glendaviesnz 

We're doing it in the JS: https://github.com/WordPress/gutenberg/pull/36293/files#diff-cd3303ad473e0bbfc9d92758c62c8f4295b31a0bb8283a3e4445c38fb73c0f4dR106

I probably should have picked up on that. 

So maybe we could have a `block-supports/utils.php` file or something with the following:

```php
/**
 * Checks whether serialization of the current block's typography properties
 * should occur.
 *
 * @param WP_Block_type $block_type  Block type.
 * @param string        $feature_set Name of block support feature set..
 * @param string        $feature     Optional name of individual feature to check.
 *
 * @return boolean Whether to serialize typography support styles & classes.
 */
function gutenberg_should_skip_block_supports_serialization( $block_type, $feature_set, $feature = null ) {
	if ( ! $block_type || ! $feature_set ) {
		return false;
	}

	$path               = array( $feature_set, '__experimentalSkipSerialization' );
	$skip_serialization = _wp_array_get( $block_type->supports, $path, false );

	if ( is_array( $skip_serialization ) ) {
		return in_array( $feature, $skip_serialization, true );
	}

	return $skip_serialization;
}
```

ramonjd replied:
> The other skip serialization methods don't have this guard? Rather than add it to all of them is it possible to make this a shared method and pass in $path as that seems to be the only difference between them?

I've done this in 9e380f0652d60833e4d8ec80fa493ad75256c76d

Thanks for the suggestion @glendaviesnz 

I smoke tested and reran all unit tests... things seem to work well, but ran out of time to retest the individual supports manually.

glendaviesnz replied:
This change looks good to me, and when tested against the group block all of the color, spacing, border, typography attributes were serialised onto the block wrapper by default, but were not when the skip serialization options were set in block.json.

ramonjd committed:
Extract `gutenberg_should_skip_block_supports_serialization` to a utils file.

tellthemachines commented on lib/block-supports/utils.php:
```diff
@@ -0,0 +1,31 @@
+ */
+
+/**
+ * Checks whether serialization of the current block's typography properties
```
Shouldn't it be "current block's style properties" or "supported properties" instead of specifically mentioning typography? Same in for the return value below.

ramonjd replied:
Good pick up, thanks for catching that copy 🍝

tellthemachines commented on lib/block-supports/elements.php:
```diff
@@ -18,6 +18,13 @@ function gutenberg_render_elements_support( $block_content, $block ) {
 	}
 
+	$block_type                    = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	$skip_link_color_serialization = gutenberg_should_skip_block_supports_serialization( $block_type, 'color', 'link' );
```
Just a thought: I'm wondering what this might look like if elements is made generic as mentioned in #31524. If we can use it with any element in any style, perhaps skipping serialization will be redundant?

ramonjd replied:
> If we can use it with any element in any style, perhaps skipping serialization will be redundant?

That's a good question, and one I can't answer, though from a superficial reading of the [supports file](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/elements.php) and the conversation of the PR you linked to, I'd probably side with you on that assumption. 😄 

I'm not sure there are any plans slated for the future of the elements mechanism 🤔 

cc @jorgefilipecosta

jorgefilipecosta replied:
> I'm not sure there are any plans slated for the future of the elements mechanism 🤔

Right now we are artificially limiting to some elements like links but the mechanism is generic and we can expand it. I'm not sure if someone is working on that.

I agree ideally the elements mechanism would remove the need to skip serialization but I guess until that mechanism is expanded we can have experimental flags to skip serializtion and then we check if we can use elements for these cases.

tellthemachines commented on lib/block-supports/layout.php:
```diff
@@ -65,7 +66,7 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 		$style .= "$selector .alignright { float: right; margin-left: 2em; margin-right: 0; }";
 		if ( $has_block_gap_support ) {
-			$gap_style = $gap_value ? $gap_value : 'var( --wp--style--block-gap )';
+			$gap_style = $gap_value && ! $should_skip_gap_serialization ? $gap_value : 'var( --wp--style--block-gap )';
```
If we're skipping serialization would we still expect there to be an output on the front end, even if it's a default value?

ramonjd replied:
Yeah, this is another tricky one. 🤣 

I agree in principle, and it did cross my mind, however there's a consideration that caused me to hesitate: `--wp--style--block-gap`'s value comes from `theme.json`.

See: https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/themes/theme-json.md#what-is-blockgap-and-how-can-i-use-it

So my reasoning at the time was, since `skipSerialization` acts at the block level, and is indeed meant to offer [finer-grained control of block supports](https://github.com/WordPress/gutenberg/issues/28913), it probably shouldn't override the theme's settings. 

Skipping serialization, as far as I understand it, skips user-facing block customization, that is the editor styles controls. 

Aside from all that, `blockGap` is another one of those anomalies. It straddles block supports and layout, and doesn't make life easy when working with both.

Hopefully we can consolidate/homogenize the use of layout and blockGap in the effort to make styles and layouts consistent. See https://github.com/WordPress/gutenberg/issues/39336

tellthemachines replied:
Yeah this is a tough one. On the whole I'm more inclined to leave it out because the theme only defines the variable value, not where to use it, and I'd expect skipSerialization to not output _anything_. But I think we can leave it as is and revisit later; outputting a couple of default styles shouldn't be a huge deal.

tellthemachines commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -304,14 +327,23 @@ export const withBlockControls = createHigherOrderComponent(
+
+		const elements = skipSerialization
+			? omit( props.attributes.style?.elements, [ 'link' ] )
+			: props.attributes.style?.elements;
```
This logic feels super awkward here. I'm wondering if we have an actual use case for skipping serialization of links. 

If we do leave this here, it would be good to have a comment explaining it as we'll probably have to remove it at some point if we change the behaviour of the elements support. Also perhaps `skipSerialization` could be renamed to something more specific like `skipLinkColorSerialization`?

ramonjd replied:
> I'm wondering if we have an actual use case for skipping serialization of links.

To be honest I don't know if there's a real-world use case. 

My only guess is that it's there for consistency for now, in case someone added

```json
		"color": {
			"text": true,
			"link": true,
			"__experimentalSkipSerialization": [ "link", "text" ]
		},
```

 for example, and was then left wondering why `text` was being skipped but `link` not. 

I'll try to formulate a comment and will change the method name. Thanks for the suggestion!

tellthemachines commented on lib/block-supports/layout.php:
```diff
@@ -28,14 +28,15 @@ function gutenberg_register_layout_support( $block_type ) {
+ * @param array   $layout                        Layout object. The one that is passed has already checked the existence of default block layout.
+ * @param boolean $has_block_gap_support         Whether the theme has support for the block gap.
+ * @param string  $gap_value                     The block gap value to apply.
+ * @param boolean $should_skip_gap_serialization Whether to skip applying the user-defined value set in the editor.
```
Should we not also allow skipping serialization for the whole layout feature too? The Navigation block already implements [custom logic](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/edit/index.js#L331) to apply the layout styles to its inner elements and blocks, so there's definitely a use case.

ramonjd replied:
> Should we not also allow skipping serialization for the whole layout feature too?

Good point! Thanks for mentioning it. I think this would be a good follow up PR. 

I'm guessing @aaronrobertshaw started the work here with a view of extending block supports (border, spacing et. al) that already use `skipSerialization` in various block.jsons today.

The layout implementation, being a little "different" might also need a level of testing that might be more suited to a dedicated PR. 😄

tellthemachines replied:
Yup, layout is complex enough that a follow-up PR sounds right.

aaronrobertshaw replied:
I'm only just getting up to speed after some recent AFK so might be missing some context.

When originally creating this I found a few block supports that simply didn't maintain/add the ability to skip serialization as per other block supports. Layout was one example of this. It also changed several times during the long life span of this PR. Early on, I had a skip for the entire layout feature but the support was reworked which, from memory, removed the code where that could easily be performed.

I think a separate PR would definitely be a good idea given how complicated layout support is and how much it's already changed in its relatively short life.

tellthemachines reviewed (commented):
This is working well! Tested skipping serialization of random features on a few blocks and all worked as expected.

I mostly left questions below to try and better understand what we're aiming for here.

ramonjd committed:
remove bad copy/pasta reference to typography

ramonjd commented:
> This is working well! Tested skipping serialization of random features on a few blocks and all worked as expected.

>  I mostly left questions below to try and better understand what we're aiming for here.

Thanks @tellthemachines I really appreciate the 👀 

I've commandeered this while @aaronrobertshaw takes a well-deserved rest, so I might not have all the context on those questions. The linked PR https://github.com/WordPress/gutenberg/pull/36104, which depends on and is the motivation for the changes in this PR, provides an example of the initial use case.

ramonjd committed:
Renamed skipSerialization method to reflect the closure context (Elements API)
Added a dev comment to warn about future changes (if any) to the Elements API

aaronrobertshaw committed:
Fix whitespace errors in spacing support php tests

aaronrobertshaw commented:
I've rebased this and fixed some whitespace errors in the spacing block support PHP unit tests.

✅ Unit tests are passing
✅ Bulk skipping of serialization still works for static blocks across various block supports
✅ Skipping individual block support features for static blocks works
✅ Bulk skipping of serialization for dynamic blocks works
✅ Skipping individual block support features for static blocks works

Primarily used the Group and Paragraph blocks for testing static blocks opting into any missing supports as needed. Used the Site Title block for testing dynamic blocks.

I believe this is looking good but it would be great to get an extra review and approval.

glendaviesnz reviewed (approved):
This tested as advertised for me with Group, Button, Paragraph and Site Title, with existing settings, blanket skipping, and with individual feature skipping
